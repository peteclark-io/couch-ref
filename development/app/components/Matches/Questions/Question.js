'use strict'

import React from 'react';
import classNames from 'classnames';
import bootstrap from 'bootstrap/dist/css/bootstrap.css';

import styles from './styles.css';

const Question = React.createClass({

   propTypes: {
      question: React.PropTypes.string,
      time: React.PropTypes.string,
      type: React.PropTypes.string,
      number: React.PropTypes.number
   },

   render: function() {
      return (
         <div>
            <h3>{this.props.time}</h3>
            <h1>{this.props.question}</h1>
            <div className={bootstrap.row}>
                <div className={bootstrap['col-xs-6']}>
                    <a className={classNames(styles['action-button'], styles.green, styles.animate, styles.shadow)}>Yes</a>
                </div>
                <div className={bootstrap['col-xs-6']}>
                    <a className={classNames(styles['action-button'], styles.red, styles.animate, styles.shadow)}>No</a>
                </div>
            </div>
         </div>
      );
   }
});

export default Question;
