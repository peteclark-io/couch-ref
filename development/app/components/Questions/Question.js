'use strict';

import React from 'react';

import {connect} from 'react-redux';
import {Link} from 'react-router';
import classNames from 'classnames';
import bootstrap from 'bootstrap/dist/css/bootstrap.css';

import {ThreeBounce} from 'better-react-spinkit';
import {saveVote} from '../../core/db-actions';

import {vote} from '../../ducks/user';

import ResultsIndicator from './ResultsIndicator';
import styles from './styles.css';
import buttons from './buttons.css';

const Question = React.createClass({

   contextTypes: {
     user: React.PropTypes.object
   },

   propTypes: {
      question: React.PropTypes.shape({
         id: React.PropTypes.string,
         question: React.PropTypes.string,
         controversial: React.PropTypes.bool,
         asked: React.PropTypes.string,
         time: React.PropTypes.string,
         description: React.PropTypes.string,
         decision: React.PropTypes.string
      }),
      votedOn: React.PropTypes.bool,
      vote: React.PropTypes.func
   },

   render: function() {
       if (!this.props.question){
         return (
           <div className={styles.loading}>
             <ThreeBounce />
           </div>
         );
       }

      return (
         <div>
            {
               this.props.question.controversial ?
               <div className={bootstrap.row}>
                  <div className={bootstrap['col-xs-12']}>
                     <h2 className={styles.controversy}>Game Changer!</h2>
                  </div>
               </div>
               : null
            }

            <div className={bootstrap.row}>
               <div className={classNames(bootstrap['col-xs-12'], bootstrap['col-sm-2'])}>
                  <h3 className={styles.time}><small>{this.props.question.time}</small></h3>
               </div>
               <div className={classNames(bootstrap['col-xs-12'], bootstrap['col-sm-10'])}>
                  <Link className={styles.link} to={`/question/${this.props.question.id}`}><h3>{this.props.question.question}</h3></Link>
                  {
                     this.props.question.decision && this.props.question.decision !== "" ?
                        <h4>REFEREE&#39;S CALL: <span className={styles.decision}>{this.props.question.decision}</span></h4>
                     : null
                  }

                  {
                     this.props.question.description && this.props.question.description !== "" ?
                        <h4><small>{this.props.question.description}</small></h4>
                     : null
                  }
               </div>
               <div className={classNames(bootstrap['col-xs-12'], bootstrap['col-sm-offset-2'], bootstrap['col-sm-10'])}>
                  {
                    !this.props.votedOn ?
                    <div>
                      <div className={styles.spacer}></div>
                      <a onClick={() => {this.props.vote(this.props.question, true)}}
                         className={classNames(buttons['action-button'], buttons['dark-blue'], buttons.animate)}>Yes</a>
                      <a onClick={() => {this.props.vote(this.props.question, false)}}
                         className={classNames(buttons['action-button'], buttons.blue, buttons.animate)}>No</a>
                    </div>
                    :
                    <ResultsIndicator id={this.props.question.id} />
                  }
               </div>
            </div>
         </div>
      );
   }
});

const getQuestion = (state = {questions: {}}, id) => {
  return state.questions[id] ? state.questions[id] : {};
};

const getVotes = (state = { user: {votes: {} } }, id) => {
  return state.user.votes && state.user.votes[id] ? true : false;
};

const mapStateToProps = (state, ownProps) => {
   return {
     question: getQuestion(state, ownProps.id),
     votedOn: getVotes(state, ownProps.id)
   };
};

const mapDispatchToProps = (dispatch) => {
  return {
    vote: (question, result) => {
      saveVote(question, result);
      dispatch(vote(question, result));
    }
  }
};

const LiveQuestion = connect(
  mapStateToProps,
  mapDispatchToProps
)(Question);

export default LiveQuestion;
