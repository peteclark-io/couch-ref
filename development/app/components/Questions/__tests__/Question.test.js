'use strict';

import React from 'react';
import {Question} from '../Question';
import {mount} from 'enzyme';
import renderer from 'react-test-renderer';

it('Show loading screen when no question.', () => {
   const rendered = mount(
      <Question />
   );
   expect(rendered.find('.loading')).toHaveLength(1);
});

it('Displays vanilla question.', () => {
   const question = {id: 'id', question: 'Is this a good test?', scored: false, time: 'Build Up'}
   const rendered = mount(
      <Question question={question} />
   );
   expect(rendered.html()).toMatchSnapshot();
});

it('Displays controversial question.', () => {
   const question = {id: 'id', question: 'Is this a good test?', controversial: true, scored: false, time: 'Build Up'}
   const rendered = mount(
      <Question question={question} />
   );
   expect(rendered.html()).toMatchSnapshot();
});

it('Displays question with referee decision.', () => {
   const question = {id: 'id', question: 'Is this a good test?', controversial: false, decision: 'Yes', scored: false, time: 'Build Up'}
   const rendered = mount(
      <Question question={question} />
   );
   expect(rendered.html()).toMatchSnapshot();
});

it('Displays question with description.', () => {
   const question = {id: 'id', question: 'Is this a good test?', controversial: true, description: 'I\'m really not sure anymore.', scored: false, time: 'Build Up'}
   const rendered = mount(
      <Question question={question} />
   );
   expect(rendered.html()).toMatchSnapshot();
});

it('Displays voting closed.', () => {
   const question = {id: 'id', question: 'Is this a good test?', controversial: true, description: 'I\'m really not sure anymore.', scored: true, time: 'Build Up'}
   const rendered = mount(
      <Question question={question} votedOn={false} />
   );
   expect(rendered.html()).toMatchSnapshot();
});
