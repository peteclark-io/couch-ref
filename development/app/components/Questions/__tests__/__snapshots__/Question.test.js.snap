exports[`test Displays controversial question. 1`] = `"<div><div class=\"row\"><div class=\"col-xs-12\"><h2 class=\"controversy\">Game Changer!</h2></div></div><div class=\"row\"><div class=\"col-xs-12 col-sm-2\"><h3 class=\"time\"><small>Build Up</small></h3></div><div class=\"col-xs-12 col-sm-10\"><h3>Is this a good test?</h3></div><div class=\"col-xs-12 col-sm-offset-2 col-sm-10\"><div><div class=\"spacer\"></div><a class=\"action-button yes animate button\">Yes</a><a class=\"action-button no animate button\">No</a></div></div></div></div>"`;

exports[`test Displays question with description. 1`] = `"<div><div class=\"row\"><div class=\"col-xs-12\"><h2 class=\"controversy\">Game Changer!</h2></div></div><div class=\"row\"><div class=\"col-xs-12 col-sm-2\"><h3 class=\"time\"><small>Build Up</small></h3></div><div class=\"col-xs-12 col-sm-10\"><h3>Is this a good test?</h3><h4><small>I\'m really not sure anymore.</small></h4></div><div class=\"col-xs-12 col-sm-offset-2 col-sm-10\"><div><div class=\"spacer\"></div><a class=\"action-button yes animate button\">Yes</a><a class=\"action-button no animate button\">No</a></div></div></div></div>"`;

exports[`test Displays question with referee decision. 1`] = `"<div><div class=\"row\"><div class=\"col-xs-12 col-sm-2\"><h3 class=\"time\"><small>Build Up</small></h3></div><div class=\"col-xs-12 col-sm-10\"><h3>Is this a good test?</h3><h4><!-- react-text: 9 -->REFEREE\'S CALL: <!-- /react-text --><span class=\"decision\">Yes</span></h4></div><div class=\"col-xs-12 col-sm-offset-2 col-sm-10\"><div><div class=\"spacer\"></div><a class=\"action-button yes animate button\">Yes</a><a class=\"action-button no animate button\">No</a></div></div></div></div>"`;

exports[`test Displays vanilla question. 1`] = `"<div><div class=\"row\"><div class=\"col-xs-12 col-sm-2\"><h3 class=\"time\"><small>Build Up</small></h3></div><div class=\"col-xs-12 col-sm-10\"><h3>Is this a good test?</h3></div><div class=\"col-xs-12 col-sm-offset-2 col-sm-10\"><div><div class=\"spacer\"></div><a class=\"action-button yes animate button\">Yes</a><a class=\"action-button no animate button\">No</a></div></div></div></div>"`;

exports[`test Displays voting closed. 1`] = `"<div><div class=\"row\"><div class=\"col-xs-12\"><h2 class=\"controversy\">Game Changer!</h2></div></div><div class=\"row\"><div class=\"col-xs-12 col-sm-2\"><h3 class=\"time\"><small>Build Up</small></h3></div><div class=\"col-xs-12 col-sm-10\"><h3>Is this a good test?</h3><h4><small>I\'m really not sure anymore.</small></h4></div><div class=\"col-xs-12 col-sm-offset-2 col-sm-10\"><h4 class=\"closed\">Voting closed!</h4></div></div></div>"`;
